include/revng/Pipeline/Structure.h

```diff
@@ -0,0 +1,224 @@
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <initializer_list>
+#include <optional>
+#include <string>
+#include <vector>
+
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/ADT/iterator_range.h"
+#include "llvm/Support/Casting.h"
+#include "llvm/Support/Error.h"
+#include "llvm/Support/raw_ostream.h"
+
+#include "revng/Pipeline/Errors.h"
+#include "revng/Pipeline/Step.h"
+
+namespace Pipeline {
+
+class PipelineExecutionEntry {
+public:
+  Step *ToExecute;
+  ContainersStatus Objectives;
+
+  PipelineExecutionEntry(Step &ToExecute, ContainersStatus Objectives) :
+    ToExecute(&ToExecute), Objectives(std::move(Objectives)) {}
+};
+
+class BoundedStep {
+private:
+  std::string PreviousStepName;
+  Step CurrentStep;
+
+public:
+  BoundedStep(std::string PreviousStepName, Step CurrentStep) :
+    PreviousStepName(std::move(PreviousStepName)),
+    CurrentStep(std::move(CurrentStep)) {}
+
+  bool hasPredecessor() const { return not PreviousStepName.empty(); }
+  const Step &getStep() const { return CurrentStep; }
+  Step &getStep() { return CurrentStep; }
+  const std::string &getPreviousStepName() const { return PreviousStepName; }
+
+  Step *operator->() { return &CurrentStep; }
+  const Step *operator->() const { return &CurrentStep; }
+
+  Step &operator*() { return CurrentStep; }
+  const Step &operator*() const { return CurrentStep; }
+};
+
+/// A pipeline structure is a sequence of steps connected one to
+/// the next, this is the level of abstraction where operations
+/// such as cross step requirements deduction can be asked.
+class Structure {
+public:
+  using Map = llvm::StringMap<BoundedStep>;
+  using iterator = revng::MapToValueIteratorType<Map::iterator>;
+  using const_iterator = revng::MapToValueIteratorType<Map::const_iterator>;
+
+private:
+  Map Steps;
+  std::vector<std::string> ReversePostOrderIndexes;
+
+public:
+  using State = llvm::StringMap<ContainersStatus>;
+
+  Step &add(llvm::StringRef PreviousName, Step &&NewStep);
+
+  BoundedStep *getPredecessor(const BoundedStep &Current) {
+    if (not Current.hasPredecessor())
+      return nullptr;
+    return &Steps.find(Current.getPreviousStepName())->second;
+  }
+
+  const BoundedStep *getPredecessor(const BoundedStep &Current) const {
+    if (not Current.hasPredecessor())
+      return nullptr;
+    return &Steps.find(Current.getPreviousStepName())->second;
+  }
+
+  Step *getPredecessor(const Step &Current) {
+    const auto &Curr = *Steps.find(Current.getName());
+    if (not Curr.second.hasPredecessor())
+      return nullptr;
+    return &Steps.find(Curr.second.getPreviousStepName())->second.getStep();
+  }
+
+  const Step *getPredecessor(const Step &Current) const {
+    const auto &Curr = *Steps.find(Current.getName());
+    if (not Curr.second.hasPredecessor())
+      return nullptr;
+    return &Steps.find(Curr.second.getPreviousStepName())->second.getStep();
+  }
```

Comment:
I'm not super happy about this string business.
I think there is a way to design this where the name just becomes a field in `Step`, all the references are `Step *` and `getPredecessor` is just a method of `Step`.

This design leaks how it was plugged on top of something that was there before. See if you can review it.